src/lightning/data/streaming/reader.py

```diff
@@ -57,42 +82,72 @@ def delete(self, chunk_indexes: List[int]) -> None:
             self._to_delete_queue.put(chunk_index)
 
     def _delete(self, chunk_index: int) -> None:
-        chunk_filepath, begin, _ = self._config[ChunkedIndex(index=-1, chunk_index=chunk_index)]
-
-        if os.path.exists(chunk_filepath):
-            os.remove(chunk_filepath)
+        """Inform the item loader of the chunk to delete."""
+        chunk_filepath, _, _ = self._config[ChunkedIndex(index=-1, chunk_index=chunk_index)]
+        self._item_loader.delete(chunk_index, chunk_filepath)
 
     def stop(self) -> None:
         """Receive the list of the chunk indices to download for the current epoch."""
         self._to_stop_queue.put(None)
 
+    def _maybe_delete_chunks(self) -> None:
+        try:
+            # Whether the reader has already finished processing a chunk
+            chunk_index = self._to_delete_queue.get(timeout=0.01)
+            self._pre_download_counter -= 1
+
+            # Store the current chunk index
+            self._chunks_index_to_be_deleted.append(chunk_index)
+
+            # Get the current cache size and decide whether we need to start cleanup. Otherwise, keep track of it
+            while (
+                self._max_cache_size
+                and self._chunks_index_to_be_deleted
+                and _get_folder_size(self._parent_cache_dir) >= self._max_cache_size
+            ):
+                # Delete the oldest chunk
+                self._delete(self._chunks_index_to_be_deleted.pop(0))
+        except Empty:
+            pass
+        except OSError as e:
+            # handle closed queue before the thread terminates
+            if "handle is closed" in str(e):
+                pass
+            else:
+                raise e
```

Comment:
The OSError handling here should go over the delete call, not everything. In addition, let's debug-log the error we are skipping?

Reply:
This would trigger only if your script exit faster than the queue is consumed and we validate on the text. I think it is safe to have it.